traffic_ops/traffic_ops_golang/api/api.go

```diff
@@ -0,0 +1,304 @@
+// This is a helper for the common case; not using this in unusual cases is perfectly acceptable.
+func WriteRespAlert(w http.ResponseWriter, r *http.Request, level tc.AlertLevel, msg string) {
+	resp := struct{ tc.Alerts }{tc.CreateAlerts(level, msg)}
+	respBts, err := json.Marshal(resp)
+	if err != nil {
+		HandleErr(w, r, http.StatusInternalServerError, nil, errors.New("marshalling JSON: "+err.Error()))
+		return
+	}
+	w.Header().Set(tc.ContentType, tc.ApplicationJson)
+	w.Write(respBts)
+}
+
+// IntParams parses integer parameters, and returns map of the given params, or an error if any integer param is not an integer. The intParams may be nil if no integer parameters are required. Note this does not check existence; if an integer paramter is required, it should be included in the requiredParams given to NewInfo.
+// This is a helper for the common case; not using this in unusual cases is perfectly acceptable.
+func IntParams(params map[string]string, intParamNames []string) (map[string]int, error) {
+	intParams := map[string]int{}
+	for _, intParam := range intParamNames {
+		valStr, ok := params[intParam]
+		if !ok {
+			continue
+		}
+		valInt, err := strconv.Atoi(valStr)
+		if err != nil {
+			return nil, errors.New("parameter '" + intParam + "'" + " not an integer")
+		}
+		intParams[intParam] = valInt
+	}
+	return intParams, nil
+}
+
+// ParamsHaveRequired checks that params have all the required parameters, and returns nil on success, or an error providing information on which params are missing.
+func ParamsHaveRequired(params map[string]string, required []string) error {
+	missing := []string{}
+	for _, requiredParam := range required {
+		if _, ok := params[requiredParam]; !ok {
+			missing = append(missing, requiredParam)
+		}
+	}
+	if len(missing) > 0 {
+		return errors.New("missing required parameters: " + strings.Join(missing, ", "))
+	}
+	return nil
+}
+
+// StripParamJSON removes ".json" trailing any parameter value, and returns the modified params.
+// This allows the API handlers to transparently accept /id.json routes, as allowed by the 1.x API.
+func StripParamJSON(params map[string]string) map[string]string {
+	for name, val := range params {
+		if strings.HasSuffix(val, ".json") {
+			params[name] = val[:len(val)-len(".json")]
+		}
+	}
+	return params
+}
+
+// AllParams takes the request (in which the router has inserted context for path parameters), and an array of parameters required to be integers, and returns the map of combined parameters, and the map of int parameters; or a user or system error and the HTTP error code. The intParams may be nil if no integer parameters are required.
+// This is a helper for the common case; not using this in unusual cases is perfectly acceptable.
+func AllParams(req *http.Request, required []string, ints []string) (map[string]string, map[string]int, error, error, int) {
```

Comment:
This is a helper function, designed to make usage easy and terse. Using a named return would add additional lines of code to the usage. The `api` package has a common idiom of returning userErr, sysErr, errCode. Changing that to structs would make usage much more verbose, require users to remember wholly unnecessary names, and defeat the purpose of the package.

>It's unclear what each of these return values mean

>returns the map of combined parameters, and the map of int parameters; or a user or system error and the HTTP error code.

Reply:
Yeah, with some careful parameter counting, I was able to eventually determine that "user or system error" were two different returns. I still think named returns would be easier to read here:

    func AllParams(req *http.Request, required []string, ints []string) (stringParams map[string]string, intParams map[string]int, userError error, systemError error, httpCode int)

And just because you name the returns doesn't mean you need to use naked returns. Variable names are just comments that a compiler can use.